File: Sample_Code/Ruby/BluePayPayment_BP10Emu.rb
#!/usr/bin/ruby
require "net/http"
require "net/https"
require "uri"
require "digest/md5"

class BluePayPayment_BP10Emu
  @@SERVER = "secure.bluepay.com"

  def initialize(account,key,mode='TEST')
    @ACCOUNT_ID = account
    @SECRET_KEY = key
    @PARAM_HASH = { 'MODE' => mode }
    @RETURN_HASH = Hash.new()
  end

  def set_param(key, val)
    @PARAM_HASH[key] = val
  end

  # Set up a credit card payment.
  def set_cc_information(account, expire, cvv='')
    @PARAM_HASH['PAYMENT_TYPE'] = 'CREDIT'
    @PARAM_HASH['CC_NUM'] = account
    @PARAM_HASH['CC_EXPIRES'] = expire
    @PARAM_HASH['CVCVV2'] = cvv
  end

  # Set up an ACH transaction.  Expects:
  # acc_type: C for Checking, S for Savings
  # routing: Bank routing number
  # account: Customer's checking or savings account number
  # doc_type: WEB, TEL, ARC, etc -- see docs.  Optional.
  # REMEMBER: Ach requires some other fields,
  # such as address and phone 

  def set_ach_information(routing, account, acc_type, doc_type='')
    @PARAM_HASH['PAYMENT_TYPE'] = 'ACH'
    @PARAM_HASH['ACH_ROUTING'] = routing
    @PARAM_HASH['ACH_ACCOUNT'] = account
    @PARAM_HASH['ACH_ACCOUNT_TYPE'] = acc_type
    @PARAM_HASH['DOC_TYPE'] = doc_type
  end

  # Set up a sale
  def sale(amount, trans_id='')
    @PARAM_HASH['TRANSACTION_TYPE'] = 'SALE'
    @PARAM_HASH['AMOUNT'] = amount
    @PARAM_HASH['RRNO'] = trans_id
  end

  # Set up an Auth
  def auth(amount, trans_id='')
    @PARAM_HASH['TRANSACTION_TYPE'] = 'AUTH'
    @PARAM_HASH['AMOUNT'] = amount
    @PARAM_HASH['RRNO'] = trans_id
  end

  # Capture an Auth
  def capture(trans_id, amount='')
    @PARAM_HASH['TRANSACTION_TYPE'] = 'CAPTURE'
    @PARAM_HASH['AMOUNT'] = amount
    @PARAM_HASH['RRNO'] = trans_id
  end

  # Refund
  def refund(trans_id, amount='')
    @PARAM_HASH['TRANSACTION_TYPE'] = 'REFUND'
    @PARAM_HASH['RRNO'] = trans_id
    @PARAM_HASH['AMOUNT'] = amount
  end

  # Void
  def void(trans_id)
    @PARAM_HASH['TRANSACTION_TYPE'] = 'VOID'
    @PARAM_HASH['AMOUNT'] = ''
    @PARAM_HASH['RRNO'] = trans_id
  end

  # Sets customer information for the transaction
  def set_customer_information(name1, name2, addr1, city, state, zip, addr2='', country='')
    @PARAM_HASH['NAME1'] = name1
    @PARAM_HASH['NAME2'] = name2
    @PARAM_HASH['ADDR1'] = addr1
    @PARAM_HASH['CITY'] = city
    @PARAM_HASH['STATE'] = state
    @PARAM_HASH['ZIPCODE'] = zip
    @PARAM_HASH['ADDR2'] = addr2
    @PARAM_HASH['COUNTRY'] = country
  end

  # Set customer phone #
  def set_phone(phone)
    @PARAM_HASH['PHONE'] = phone
  end

  # Set customer E-mail address
  def set_email(email)
    @PARAM_HASH['EMAIL'] = email
  end

  # Set MEMO field
  def set_memo(memo)
    @PARAM_HASH['COMMENT'] = memo
  end

  # Set CUSTOM_ID field
  def set_custom_id1(custom_id1)
    @PARAM_HASH['CUSTOM_ID'] = custom_id1
  end

  # Set CUSTOM_ID2 field
  def set_custom_id2(custom_id2)
    @PARAM_HASH['CUSTOM_ID2'] = custom_id2
  end

  # Set INVOICE_ID field
  def set_invoice_id(invoice_id)
    @PARAM_HASH['INVOICE_ID'] = invoice_id
  end

  # Set ORDER_ID field
  def set_order_id(order_id)
    @PARAM_HASH['ORDER_ID'] = order_id
  end

  # Set AMOUNT_TIP field
  def set_amount_tip(amount_tip)
    @PARAM_HASH['AMOUNT_TIP'] = amount_tip
  end

  # Set AMOUNT_TAX field
  def set_amount_tax(amount_tax)
    @PARAM_HASH['AMOUNT_TAX'] = amount_tax
  end

  # Set AMOUNT_FOOD field
  def set_amount_food(amount_food)
    @PARAM_HASH['AMOUNT_FOOD'] = amount_food
  end

  # Set AMOUNT_MISC field
  def set_amount_misc(amount_misc)
    @PARAM_HASH['AMOUNT_MISC'] = amount_misc
  end

  # Set fields for a recurring payment
  def add_recurring_fields(rebill_first_date, rebill_expr, reb_cycles, reb_amount)
    @PARAM_HASH['REBILLING'] = '1'
    @PARAM_HASH['REB_FIRST_DATE'] = rebill_first_date
    @PARAM_HASH['REB_EXPR'] = rebill_expr
    @PARAM_HASH['REB_CYCLES'] = reb_cycles
    @PARAM_HASH['REB_AMOUNT'] = reb_amount
  end

  # Set fields to do an update on an existing rebilling cycle
  def update_rebilling_cycle(rebill_id, rebill_next_date, rebill_expr, rebill_cycles, rebill_amount,
    rebill_next_amount)
    @PARAM_HASH['TRANS_TYPE'] = "SET"
    @PARAM_HASH['REBILL_ID'] = rebill_id
    @PARAM_HASH['NEXT_DATE'] = rebill_next_date
    @PARAM_HASH['REB_EXPR'] = rebill_expr
    @PARAM_HASH['REB_CYCLES'] = rebill_cycles
    @PARAM_HASH['REB_AMOUNT'] = rebill_amount
    @PARAM_HASH['NEXT_AMOUNT'] = rebill_next_amount
  end

  # Set fields to cancel an existing rebilling cycle
  def cancel_rebilling_cycle(rebill_id)
    @PARAM_HASH["TRANS_TYPE"] = "SET"
    @PARAM_HASH["STATUS"] = "stopped"
    @PARAM_HASH["REBILL_ID"] = rebill_id
  end

  # Set fields to get the status of an existing rebilling cycle
  def get_rebilling_cycle_status(rebill_id)
    @PARAM_HASH["TRANS_TYPE"] = "GET"
    @PARAM_HASH["REBILL_ID"] = rebill_id
  end

  # Updates an existing rebilling cycle's payment information.   
  def update_rebilling_payment_information(template_id)
    @PARAM_HASH["TEMPLATE_ID"] = template_id
  end

  # Gets a report on all transactions within a specified date range
  def get_transaction_report(report_start, report_end, subaccounts_searched, do_not_escape = '', errors = '')
    @PARAM_HASH["QUERY_BY_SETTLEMENT"] = '0'
    @PARAM_HASH["REPORT_START_DATE"] = report_start
    @PARAM_HASH["REPORT_END_DATE"] = report_end
    @PARAM_HASH["QUERY_BY_HIERARCHY"] = subaccounts_searched
    @PARAM_HASH["DO_NOT_ESCAPE"] = do_not_escape
    @PARAM_HASH["EXCLUDE_ERRORS"] = errors
  end

  # Gets a report on all settled transactions within a specified date range
  def get_settled_transaction_report(report_start, report_end, subaccounts_searched, do_not_escape = '', errors = '')
    @PARAM_HASH["QUERY_BY_SETTLEMENT"] = '1'
    @PARAM_HASH["REPORT_START_DATE"] = report_start
    @PARAM_HASH["REPORT_END_DATE"] = report_end
    @PARAM_HASH["QUERY_BY_HIERARCHY"] = subaccounts_searched
    @PARAM_HASH["DO_NOT_ESCAPE"] = do_not_escape
    @PARAM_HASH["EXCLUDE_ERRORS"] = errors
  end

  # Gets data on a specific transaction
  def get_single_trans_query(report_start, report_end, errors = "")
    @PARAM_HASH["REPORT_START_DATE"] = report_start
    @PARAM_HASH["REPORT_END_DATE"] = report_end
    @PARAM_HASH["EXCLUDE_ERRORS"] = errors
  end

  # Queries by a specific Transaction ID. To be used with get_single_trans_query
  def query_by_transaction_id(trans_id)
    @PARAM_HASH["id"] = trans_id;
  end

  # Queries by a specific Payment Type. To be used with get_single_trans_query
  def query_by_payment_type(pay_type)
    @PARAM_HASH["payment_type"] = payment_type
  end

  # Queries by a specific Transaction Type. To be used with get_single_trans_query
  def query_by_trans_type(trans_type)
    @PARAM_HASH["trans_type"] = trans_type
  end

  # Queries by a specific Transaction Amount. To be used with get_single_trans_query
  def query_by_amount(amount)
    @PARAM_HASH["amount"] = amount
  end

  # Queries by a specific First Name. To be used with get_single_trans_query
  def query_by_name1(name1)
    @PARAM_HASH["name1"] = name1
  end

  # Queries by a specific Last Name. To be used with get_single_trans_query
  def query_by_name2(name2) 
    @PARAM_HASH["name2"] = name2
  end

  # Turns a hash into a nvp style string
  def uri_query(h)
    a = Array.new()
    h.each_pair {|key, val| a.push(URI.escape(key) + "=" + URI.escape(val)) }
    return a.join("&")
  end

  # Sets TAMPER_PROOF_SEAL in @PARAM_HASH
  def calc_tps()
    @PARAM_HASH["TAMPER_PROOF_SEAL"] = Digest::MD5.hexdigest(@SECRET_KEY + @ACCOUNT_ID +
      (@PARAM_HASH["TRANSACTION_TYPE"] || '') + @PARAM_HASH["AMOUNT"] + (@PARAM_HASH["REBILLING"] || '') +
      (@PARAM_HASH["REB_FIRST_DATE"] || '') + (@PARAM_HASH["REB_EXPR"] || '') + (@PARAM_HASH["REB_CYCLES"] || '') +
      (@PARAM_HASH["REB_AMOUNT"] || '') + (@PARAM_HASH["RRNO"] || '') + @PARAM_HASH["MODE"])
  end

  def calc_rebill_tps()
    @PARAM_HASH["TAMPER_PROOF_SEAL"] = Digest::MD5.hexdigest(@SECRET_KEY + @ACCOUNT_ID +
      @PARAM_HASH["TRANS_TYPE"] + @PARAM_HASH["REBILL_ID"])
  end

  def calc_report_tps()
    @PARAM_HASH["TAMPER_PROOF_SEAL"] = Digest::MD5.hexdigest(@SECRET_KEY + @ACCOUNT_ID + 
      @PARAM_HASH["REPORT_START_DATE"] + @PARAM_HASH["REPORT_END_DATE"])
  end
 
  def calc_trans_notify_tps(secret_key, trans_id, trans_status, trans_type, amount, batch_id, batch_status,
    total_count, total_amount, batch_upload_id, rebill_id, rebill_amount, rebill_status)
    @PARAM_HASH["TAMPER_PROOF_SEAL"] = Digest::MD5.hexdigest(@SECRET_KEY + trans_id + trans_status + transtype + 
      amount + batch_id + batch_status + total_count + total_amount + batch_upload_id + rebill_id + rebill_amount + rebill_status)
  end

  def process()
    ua = Net::HTTP.new(@@SERVER, 443)
    ua.use_ssl = true
    begin
    	@PARAM_HASH["REMOTE_IP"] = request.env['REMOTE_ADDR']
    rescue Exception
    end
    # Generate the query string and headers
    if (@PARAM_HASH.has_key?("QUERY_BY_HIERARCHY"))
      calc_report_tps()
      path = "/interfaces/bpdailyreport2"
      query = "ACCOUNT_ID=#{@ACCOUNT_ID}&"
      query += uri_query(@PARAM_HASH)
    elsif (@PARAM_HASH.has_key?("REPORT_START_DATE"))
      calc_report_tps()
      path = "/interfaces/stq"
      query = "ACCOUNT_ID=#{@ACCOUNT_ID}&"
      query += uri_query(@PARAM_HASH)
    elsif (@PARAM_HASH["TRANS_TYPE"] != "SET" and @PARAM_HASH["TRANS_TYPE"] != "GET")
      calc_tps()
      path = "/interfaces/bp10emu"
      query = "MERCHANT=#{@ACCOUNT_ID}&"
      query += uri_query(@PARAM_HASH)
    else
      calc_rebill_tps()
      path = "/interfaces/bp20rebadmin"
      query = "ACCOUNT_ID=#{@ACCOUNT_ID}&"
      query += uri_query(@PARAM_HASH)
    end

    queryheaders = {
      'User-Agent' => 'Bluepay Ruby Client',
      'Content-Type' => 'application/x-www-form-urlencoded'
    }
    @PARAM_HASH["VERSION"] = '1'
    # Post parameters to BluePay gateway
    headers, body = ua.post(path, query, queryheaders)
    # Split the response into the response hash.
    @RESPONSE_HASH = {}
    if path == "/interfaces/bp10emu"
      response = headers["Location"].split("?")[1]
    else
      response = headers.body
    end
    response.split("&").each { |pair| 
    (key, val) = pair.split("=")
    val = "" if(val == nil)
    @RESPONSE_HASH[URI.unescape(key)] = URI.unescape(val) 
    }
  end

  def get_response()
    return @RESPONSE_HASH
  end

  # Returns ! on WTF, E for Error, 1 for Approved, 0 for Decline
  def get_status()
    return @RESPONSE_HASH['Result']
  end

  # Returns the human-readable response from Bluepay.
  # Or a nasty error.
  def get_message()
    m = @RESPONSE_HASH['MESSAGE']
    if (m == nil or m == "")
      return "ERROR - NO MESSAGE FROM BLUEPAY"
    end
    return m
  end

  # Returns the single-character AVS response from the 
  # Card Issuing Bank
  def get_avs_code()
    return @RESPONSE_HASH['AVS']
  end

  # Same as avs_code, but for CVV2

  def get_cvv2_code()
    return @RESPONSE_HASH['CVV2']
  end

  # In the case of an approved transaction, contains the
  # 6-character authorization code from the processing network.
  # In the case of a decline or error, the contents may be junk.
  def get_auth_code()
    return @RESPONSE_HASH['AUTH_CODE']
  end

  # The all-important transaction ID.
  def get_trans_id()
    return @RESPONSE_HASH['RRNO']
  end

  # If you set up a rebilling, this'll get it's ID.
  def get_rebill_id()
    return @RESPONSE_HASH['REBID']
  end

  # masked credit card or ACH account
  def get_masked_account()
    return @RESPONSE_HASH['PAYMENT_ACCOUNT']
  end

  # card type used in transaction
  def get_card_type()
    return @RESPONSE_HASH['CARD_TYPE']
  end

  # bank account used in transaction
  def get_bank_name()
    return @RESPONSE_HASH['BANK_NAME']
  end

  def get_reb_id()
    return @RESPONSE_HASH['rebill_id']
  end

  def get_template_id()
    return @RESPONSE_HASH['template_id']
  end

  def get_rebill_status()
    return @RESPONSE_HASH['status']
  end

  def get_creation_date()
    return @RESPONSE_HASH['creation_date']
  end
  
  def get_next_date()
      return @RESPONSE_HASH['next_date']
  end

  def get_last_date()
    return @RESPONSE_HASH['last_date']
  end

  def get_sched_expression()
    return @RESPONSE_HASH['sched_expr']
  end
  
  def get_cycles_remaining()
    return @RESPONSE_HASH['cycles_remain']
  end

  def get_rebill_amount()
    return @RESPONSE_HASH['reb_amount']
  end

  def get_next_amount()
    return @RESPONSE_HASH['next_amount']
  end
 
  def get_id()
    return @RESPONSE_HASH['id']
  end 

  def get_name1()
    return @RESPONSE_HASH['name1']
  end
 
  def get_name2()
    return @RESPONSE_HASH['name2']
  end
  
  def get_payment_type()
    return @RESPONSE_HASH['payment_type']
  end

  def get_trans_type()
    return @RESPONSE_HASH['trans_type']
  end
  
  def get_amount()
    return @RESPONSE_HASH['amount']
  end

end

File: Sample_Code/Python/BluePayPayment_BP10Emu.py
import urllib
from urllib2 import Request, urlopen, URLError, HTTPError
import urlparse
import hashlib
import cgi
import os
import re

class BluePayPayment_BP10Emu:
    # Merchant fields
    accountID = ''
    secretKey = '' 
    mode = ''

    # Transaction fields
    transType = ''
    paymentType = ''
    amount = ''
    cardNumber = ''
    cvv2 = ''
    cardExpire = ''
    routingNumber = ''
    accountNumber = ''
    accountType = ''
    docType = ''

    # Customer fields
    name1 = ''
    name2 = ''
    addr1 = ''
    addr2 = ''
    city = ''
    state = ''
    zipcode = ''
    country = ''
    phone = ''
    email = ''

    # Optional fields
    memo = ''
    customID1 = ''
    customID2 = ''
    invoiceID = ''
    orderID = ''
    amountTax = ''
    amountTip = ''
    amountFood = ''
    amountMisc = ''

    # Rebilling fields
    doRebill = ''
    rebFirstDate = ''
    rebExpr = ''
    rebCycles = ''
    rebAmount = ''
    rebNextDate = ''
    rebNextAmount = ''
    templateID = ''

    # Reporting fields
    reportStartDate = ''
    reportEndDate = ''
    queryBySettlement = ''
    subaccountsSearched = ''
    doNotEscape = ''
    excludeErrors = ''
    
    # Response fields
    rebStatus = ''
    rebillID = ''
    masterID = ''
    data = ''

    url = ''

    # Class constructor. Accepts:
    # accID : Merchant's Account ID
    # secretKey : Merchant's Secret Key
    # mode : Transaction mode of either LIVE or TEST (default)
    def __init__(self, accID, secretKey, mode):
        self.accountID = accID
	self.secretKey = secretKey
	self.mode = mode

    # Performs a SALE
    def sale(self, amount, masterID=None):
        """ 
        Send a Sale request to the BluePay gateway.
        """
        self.transType = 'SALE'
        self.amount = amount
	if masterID is not None:
		self.masterID = masterID
        response = self.postData()
        return self.parseResponse(response)
 
    # Performs an AUTH
    def auth(self, amount, masterID=None):
        """ 
        Send an Auth request to the BluePay gateway
		"""
	self.transType = 'AUTH'
	self.amount = amount
	if masterID is not None:
		self.masterID = masterID
        response = self.postData() 
        return self.parseResponse(response)
    
    # Performs a CAPTURE
    def capture(self, masterID, amount=None):
        """ 
        Send a Capture request to the BluePay gateway 
        """
	self.transType = 'CAPTURE'
	self.masterID = masterID
	if amount is not None:
		self.amount = amount
        response = self.postData() 
        return self.parseResponse(response)

    # Performs a REFUND
    def refund(self, masterID, amount=None):
        """ 
        Send a Refund request to the BluePay gateway.
        """
        self.transType = 'REFUND'
	self.masterID = masterID
	if amount is not None:
            self.amount = amount
        response = self.postData()
        return self.parseResponse(response)

    # Performs a VOID
    def void(self, masterID):
        """ 
        Send a Void request to the BluePay gateway.
        """
        self.transType = 'VOID'
	self.masterID = masterID
        response = self.postData()
        return self.parseResponse(response)

    # Passes customer information into the transaction
    def setCustomerInformation(self, name1, name2, addr1, city, state,
	zipcode, addr2=None, country=None):
	self.name1 = name1
	self.name2 = name2
	self.addr1 = addr1
	self.city = city
	self.state = state
	self.zipcode = zipcode
	self.addr2 = addr2
	self.country = country
        return None
  
    # Passes credit card information into the transaction
    def setCCInformation(self, cardNum, cardExpire, cvv2=None):
	self.paymentType = 'CREDIT'
	self.cardNumber = cardNum
	self.cardExpire = cardExpire
	self.cvv2 = cvv2
	return None

    # Passes ACH information into the transaction
    def setACHInformation(self, routingNum, accNum, accType, docType=None):
	self.paymentType = 'ACH'
	self.routingNumber = routingNum
	self.accountNumber = accNum
	self.accountType = accType
	self.docType = docType
	return None
       
    # Passes rebilling information into the transaction
    def setRebillingInformation(self, firstDate, expr, cycles, amount):
	self.doRebill = '1'
	self.rebFirstDate = firstDate
	self.rebExpr = expr
	self.rebCycles = cycles
	self.rebAmount = amount
	return None

    # Passes rebilling information for a rebill update
    def updateRebill(self, rebillID, nextDate, expr, cycles, rebillAmount, nextAmount):
	self.transType = 'SET'
	self.rebillID = rebillID
	self.rebNextDate = nextDate
	self.rebExpr = expr
	self.rebCycles = cycles
	self.rebAmount = rebillAmount
	self.rebNextAmount = nextAmount
	response = self.postData()
        return self.parseResponse(response)

    # Passes rebilling information for a rebill cancel
    def cancelRebill(self, rebillID):
	self.transType = 'SET'
	self.rebStatus = 'stopped'
	self.rebillID = rebillID
	response = self.postData()
        return self.parseResponse(response)

    # Set fields to get the status of an existing rebilling cycle
    def getRebillingCycleStatus(self, rebillID):
    	self.transType = 'GET'
    	self.rebillID = rebillID
	response = self.postData()
        return self.parseResponse(response)

    # Updates an existing rebilling cycle's payment information.   
    def updateRebillingPaymentInformation(self, templateID):
    	self.templateID = templateID
	return None

    # Passes values for a call to the bpdailyreport2 API to get all transactions based on start/end dates
    def getTransactionReport(self, reportStart, reportEnd, subaccountsSearched, doNotEscape=None, errors=None):
	self.queryBySettlement = '0'
	self.reportStartDate = reportStart
	self.reportEndDate = reportEnd
	self.subaccountsSearched = subaccountsSearched
	if doNotEscape is not None:
		self.doNotEscape = doNotEscape
	if errors is not None:
		self.excludeErrors = errors
	response = self.postData()
	return response

    # Passes values for a call to the bpdailyreport2 API to get settled transactions based on start/end dates
    def getSettledTransactionReport(self, reportStart, reportEnd, subaccountsSearched, doNotEscape=None, errors=None):
	self.queryBySettlement = '1'
        self.reportStartDate = reportStart
        self.reportEndDate = reportEnd
        self.subaccountsSearched = subaccountsSearched
	if doNotEscape is not None:
            self.doNotEscape = doNotEscape
        if errors is not None:
            self.excludeErrors = errors
	response = self.postData()
	return response
 
    # Passes values for a call to the stq API to get information on a single transaction
    def getSingleTransQuery(self, reportStart, reportEnd, errors=None):
	self.reportStartDate = reportStart
	self.reportEndDate = reportEnd
	self.excludeErrors = errors
	response = self.postData()
	return response

    # Queries transactions by a specific Transaction ID. Must be used with getSingleTransQuery
    def queryByTransactionID(self, transID):
	self.transID = transID
	return None

    # Queries transactions by a specific Payment Type. Must be used with getSingleTransQuery
    def queryByPaymentType(self, payType):
	self.paymentType = payType
	return None

    # Queries transactions by a specific Transaction Type. Must be used with getSingleTransQuery
    def queryByTransType(self, transType):
	self.transType = transType
	return None

    # Queries transactions by a specific Transaction Amount. Must be used with getSingleTransQuery
    def queryByAmount(self, amount):
	self.amount = amount
	return None

    # Queries transactions by a specific First Name. Must be used with getSingleTransQuery
    def queryByName1(self, name1):
	self.name1 = name1
	return None

    # Queries transactions by a specific Last Name. Must be used with getSingleTransQuery
    def queryByName2(self, name2):
	self.name2 = name2
	return None

    # passes value into PHONE field
    def setPhone(self, phone):
	self.phone = phone
	return None

    # passes value into EMAIL field
    def setEmail(self, email):
	self.email = email
	return None

    # Passes value into CUSTOM_ID field
    def setCustomID1(self, customID1):
	self.customID1 = customID1
	return None

    # Passes value into CUSTOM_ID2 field
    def setCustomID2(self, customID2):
        self.customID2 = customID2
        return None

    # Passes value into INVOICE_ID field
    def setInvoiceID(self, invoiceID):
        self.invoiceID = invoiceID
        return None
  
    # Passes value into ORDER_ID field
    def setOrderID(self, orderID):
        self.orderID = orderID
        return None

    # Passes value into MEMO field
    def setMemo(self, memo):
        self.memo = memo
        return None

    # Passes value into AMOUNT_TAX field
    def setAmountTax(self, amountTax):
        self.amountTax = amountTax
        return None

    # Passes value into AMOUNT_TIP field
    def setAmountTip(self, amountTip):
        self.amountTip = amountTip
        return None

    # Passes value into AMOUNT_FOOD field
    def setAmountFood(self, amountFood):
        self.amountFood = amountFood
        return None

    # Passes value into AMOUNT_MISC field
    def setAmountMisc(self, amountMisc):
        self.amountMisc = amountMisc
        return None
 
    def postData(self, card=None, customer=None, order=None):
        fields = {
	    'MODE': self.mode,
	    'RRNO': self.masterID
	}
	if self.subaccountsSearched != '':
	    self.url = 'https://secure.bluepay.com/interfaces/bpdailyreport2'
	    fields.update({
		'ACCOUNT_ID': self.accountID,
		'REPORT_START_DATE' : self.reportStartDate,
		'REPORT_END_DATE' : self.reportEndDate,
		'TAMPER_PROOF_SEAL' : self.calcReportTPS(),
		'DO_NOT_ESCAPE' : self.doNotEscape,
		'QUERY_BY_SETTLEMENT' : self.queryBySettlement,
		'QUERY_BY_HIERARCHY' : self.subaccountsSearched,
		'EXCLUDE_ERRORS' : self.excludeErrors
	    })
	elif self.reportStartDate != '':
	    self.url = 'https://secure.bluepay.com/interfaces/stq'
	    fields.update({
		'ACCOUNT_ID': self.accountID,
		'REPORT_START_DATE' : self.reportStartDate,
		'REPORT_END_DATE' : self.reportEndDate,
		'TAMPER_PROOF_SEAL' : self.calcReportTPS(),
		'EXCLUDE_ERRORS' : self.excludeErrors,
		'IGNORE_NULL_STR' : '1',
		'id' : self.transID,
		'payment_type' : self.paymentType,
		'trans_type' : self.transType,
		'amount' : self.amount,
		'name1' : self.name1,
		'name2' : self.name2
	    })
	elif self.transType != 'SET' and self.transType != 'GET':
	    self.url = 'https://secure.bluepay.com/interfaces/bp10emu'
	    fields.update({
		'MERCHANT': self.accountID,
		'TRANSACTION_TYPE': self.transType,
        	'PAYMENT_TYPE': self.paymentType,
	        'AMOUNT': self.amount,
		'NAME1': self.name1,
		'NAME2': self.name2,
		'ADDR1': self.addr1,
		'ADDR2': self.addr2,
		'CITY': self.city,
		'STATE': self.state,
		'ZIPCODE': self.zipcode,
		'COUNTRY': self.country,
		'EMAIL': self.email,
		'PHONE': self.phone,
		'CUSTOM_ID': self.customID1,
		'CUSTOM_ID2': self.customID2,
		'INVOICE_ID': self.invoiceID,
		'ORDER_ID': self.orderID,
		'COMMENT': self.memo,
		'AMOUNT_TAX': self.amountTax,
		'AMOUNT_TIP': self.amountTip,
		'AMOUNT_FOOD': self.amountFood,
		'AMOUNT_MISC': self.amountMisc,
	  	'REBILLING': self.doRebill,
		'REB_FIRST_DATE': self.rebFirstDate,
		'REB_EXPR': self.rebExpr,
		'REB_CYCLES': self.rebCycles,
		'REB_AMOUNT': self.rebAmount,
	        'TAMPER_PROOF_SEAL': self.calcTPS()
	    })
	    try:
		fields.update({
		    'REMOTE_IP' : cgi.escape(os.environ["REMOTE_ADDR"])
		})
	    except KeyError:
		pass
	    if self.paymentType == 'CREDIT':
 	        fields.update({
		    'CC_NUM': self.cardNumber, 
            	    'CC_EXPIRES': self.cardExpire,
	            'CVCVV2': self.cvv2
	  	})
	    else:
		fields.update({
		    'ACH_ROUTING': self.routingNumber,
		    'ACH_ACCOUNT': self.accountNumber,
		    'ACH_ACCOUNT_TYPE': self.accountType,
		    'DOC_TYPE': self.docType
		})
	else:
	    self.url = 'https://secure.bluepay.com/interfaces/bp20rebadmin';
	    fields.update({
		'ACCOUNT_ID': self.accountID,
		'TRANS_TYPE': self.transType,
		'REBILL_ID': self.rebillID,
		'TEMPLATE_ID' : self.templateID,
		'NEXT_DATE': self.rebNextDate,
		'REB_EXPR': self.rebExpr,
		'REB_CYCLES': self.rebCycles,
		'REB_AMOUNT': self.rebAmount,
		'NEXT_AMOUNT': self.rebNextAmount,
		'STATUS': self.rebStatus,
		'TAMPER_PROOF_SEAL': self.calcRebillTPS()
		})
	response = self.request(self.url,self.createPostString(fields))
        return response

    def createPostString(self, fields):
        fields = dict([k,str(v).replace(',', '')] for (k,v) in fields.iteritems())       
        return urllib.urlencode(fields)
    
    def request(self, url, data):
        """
        Submits an https request to BluePay.
        """
        response = self.send(data)
        return response

    def send(self, data):
        """
        Send an https request.
        """
	try:
	    r = urlopen(self.url, data)
	    response = r.read()
	    return response
	except HTTPError, e:
	    if re.match("https://secure.bluepay.com/interfaces/wlcatch", e.geturl()):
		response = e.geturl()
		return response
	    return e.read()
	return "ERROR"
 
    def parseResponse(self, response):
	if self.url != 'https://secure.bluepay.com/interfaces/bp10emu':
	    response = cgi.parse_qs(response)
	else:
	    queryString = urlparse.urlparse(response)
	    response = urlparse.parse_qs(queryString.query)
        return response
    
    def __str__(self):
        return 'BluePay Python sample code for the BP10Emu API'
    
    # Functions for calculating the TAMPER_PROOF_SEAL
    def calcTPS(self):
	tpsString = (self.secretKey + self.accountID + self.transType + self.amount +
	    self.doRebill + self.rebFirstDate + self.rebExpr + self.rebCycles + 
	    self.rebAmount + self.masterID + self.mode)
	m = hashlib.md5()
	m.update(tpsString)
	return m.hexdigest()

    def calcRebillTPS(self):
        tpsString = (self.secretKey + self.accountID + self.transType + self.rebillID)
        m = hashlib.md5()
        m.update(tpsString)
        return m.hexdigest()
		
    def calcReportTPS(self):
	tpsString = (self.secretKey + self.accountID + self.reportStartDate + self.reportEndDate)
	m = hashlib.md5()
	m.update(tpsString)
	return m.hexdigest()
		
    def calcTransNotifyTPS(self):
	tpsString = (secretKey, transID, transStatus, transType, amount, batchID, batchStatus,
	    totalCount, totalAmount, batchUploadID, rebillID, rebillAmount, rebillStatus)
	m = hashlib.md5()
	m.update(tpsString)
	return m.hexdigest()
